Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,215 @@
+allocate and register an ancillary_device for each physical function on the NIC.
+An RDMA driver registers an ancillary_driver that will be matched with and
+probed for each of these ancillary_devices.
+
+Another usage case is for a PCI device whose functionality is complex enough
+that it warrants dividing it into multiple driver modules.  The ancillary_device
+and ancillary_driver matching and probing would allow the main driver to expose
+a shared memory object(s) with the peer modules to allow them to perform their
+desired function.
+
+The ancillary bus is to be used when a driver and one or more kernel modules,
+who share a common header file with the driver, need a mechanism to connect and
+provide access to a shared object allocated by the ancillary_device's
+registering driver.  The registering driver for the ancillary_device(s) and the
+kernel module(s) registering ancillary_drivers can be from the same subsystem,
+or from multiple subsystems.
+
+The emphasis here is on a common generic interface that keeps subsystem
+customization out of the bus infrastructure.
+
+When Should the Ancillary Bus Not Be Used
+=========================================
+
+If there is a need for a more specifically customized bus that provides
+functionality beyond just the matching up of client devices and client drivers,
+and sacrifices common generality to extend the bus functionality into subsystem
+specific functions, then the ancillary bus should not be used.  A custom bus
+solution, or some other method should be considered in this use case.
+
+Ancillary Devices
+=================
+
+An ancillary_device is created and registered to represent a part of its parent
+device's functionality. It is given a match_name that is used for driver
+binding and a release callback that is invoked when the device is unregistered.
+
+.. code-block:: c
+
+	struct ancillary_device {
+		struct device dev;
+		const char *match_name;
+		void (*release)(struct ancillary_device *);
+		u32 id;
+	};
+
+The ancillary_device is enumerated when it is attached to the bus. The device
+is assigned a unique ID automatically that will be appended to its name. If
+two ancillary_devices both named "foo" are registered onto the bus, they will
+have the device names, "foo.x" and "foo.y", where x and y are unique integers.
+
+Ancillary Device Memory Model and Lifespan
+------------------------------------------
+
+When a kernel driver registers an ancillary_device on the ancillary bus, we will
+use the nomenclature to refer to this kernel driver as a registering driver.  It
+is the entity that will allocate memory for the ancillary_device and register it
+on the ancillary bus.
+
+A parent object, defined in the shared header file, will contain the
+ancillary_device.  It will also contain a pointer to the shared object(s), which
+will also be defined in the shared header.  Both the parent object and the
+shared object(s) will be allocated by the ancillary_device's registering driver.
```

Comment:
We already explained what a registering driver is. Just use registering driver instead of ancillary_device's registering driver?